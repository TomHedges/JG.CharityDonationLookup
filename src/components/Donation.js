import React, { Component } from "react";
import { hot } from "react-hot-loader";
import PropTypes from "prop-types";
import { nullLiteral } from "@babel/types";
//import unknown_avatar from "./../images/unknown_avatar.svg";

function convertDate(textDate) {
  const donation_date_millis = parseInt(textDate.slice(6, textDate.length - 7));
  const donation_date = new Date(donation_date_millis);
  //const current_date = new Date();
  //const current_date_millis = current_date.getTime();
  //const date_diff = current_date_millis - donation_date_millis;

  //return date.getDate() + "/" + date.getMonth() + "/" + date.getFullYear();
  return (
    //donation_date.toLocaleDateString() + "|" + current_date.toLocaleDateString()
    donation_date.toLocaleDateString()
  );
}

function getCurrencySymbol(donation_currency) {
  if (donation_currency === "GBP") {
    return "Â£";
  } else if (donation_currency === "USD") {
    return "$";
  } else {
    return donation_currency;
  }
}

function getTaxReclaim(donation_tax_reclaim, donation_currency) {
  if (donation_tax_reclaim && donation_tax_reclaim > 0) {
    return (
      <span className="tax_reclaim">
        + {getCurrencySymbol(donation_currency)}
        {donation_tax_reclaim.toFixed(2)} Gift Aid
      </span>
    );
  } else {
    return null;
  }
}

function getDonation(
  donation_amount,
  donation_currency,
  donation_amount_local,
  donation_currency_local
) {
  if (donation_amount && donation_amount > 0) {
    if (donation_currency_local === "GBP") {
      return getCurrencySymbol(donation_currency) + donation_amount.toFixed(2);
    } else {
      return (
        getCurrencySymbol(donation_currency_local) +
        donation_amount_local.toFixed(2)
      );
    }
  } else {
    return null;
  }
}

const Donation = props => (
  <div className="donation_wrapper">
    <div className="donation_column_left">
      <img className="donor_avatar" src={props.donor_avatar_url} />
    </div>
    <div className="donation_column_right">
      <p className="donation_date">{convertDate(props.donation_date)}</p>
      <p className="donor_name">{props.donor_name}</p>
      <p>{props.donor_comment}</p>
      <p className="donation_amount">
        {getDonation(
          props.donation_amount,
          props.donation_currency,
          props.donation_amount_local,
          props.donation_currency_local
        )}
        {getTaxReclaim(props.donation_tax_reclaim, props.donation_currency)}
      </p>
    </div>
  </div>
);

Donation.propTypes = {
  donor_name: PropTypes.string.isRequired,
  donation_date: PropTypes.string.isRequired,
  donor_comment: PropTypes.string.isRequired,
  donation_amount: PropTypes.number.isRequired,
  donation_tax_reclaim: PropTypes.number.isRequired,
  donation_currency: PropTypes.string.isRequired,
  donation_currency_local: PropTypes.string.isRequired,
  donation_amount_local: PropTypes.number.isRequired,
  donor_avatar_url: PropTypes.string.isRequired
};

export default hot(module)(Donation);
