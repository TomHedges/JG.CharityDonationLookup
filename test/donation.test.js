import Donation from "../src/components/Donation";

/**
 * Unit tests for the Donation component
 */

const DONOR_NAME = "Tom Hedges";
const DONATION_DATE = "/Date(1554076800000+0000)/";
const DONATION_DATE_DISPLAY_FORMAT = "01/04/2019";
const DONOR_COMMENT = "Good luck!";
const DONATION_AMOUNT = 10;
const DONATION_CURRENCY = "GBP";
const DONATION_AMOUNT_CONVERTED = "£10.00";
const DONATION_TAX_RECLAIM = 2.5;
const DONATION_AMOUNT_LOCAL = 10;
const DONATION_CURRENCY_LOCAL = "GBP";
const DONATION_TAX_RECLAIM_CONVERTED = "+ £2.50 Gift Aid";
const DONOR_AVATAR_URL =
  "https://images.justgiving.com/image/f1e9879a-dace-443a-a453-d0636e66a348.jpg?template=profilesummary";
const NO_TAX_RECLAIM = 0;
const USD_DONATION_AMOUNT_LOCAL = 15;
const USD_DONATION_CURRENCY_LOCAL = "USD";
const USD_DONATION_AMOUNT_CONVERTED = "$15.00";
const DONATION_KEY =
  DONOR_NAME + DONATION_DATE + DONOR_COMMENT + DONATION_AMOUNT;

function createTestProps(props) {
  return {
    donor_name: DONOR_NAME,
    donation_date: DONATION_DATE,
    donor_comment: DONOR_COMMENT,
    donation_amount: DONATION_AMOUNT,
    donation_currency: DONATION_CURRENCY,
    donation_tax_reclaim: DONATION_TAX_RECLAIM,
    donation_amount_local: DONATION_AMOUNT_LOCAL,
    donation_currency_local: DONATION_CURRENCY_LOCAL,
    donor_avatar_url: DONOR_AVATAR_URL,
    donation_key: DONATION_KEY,
    ...props
  };
}

let wrapper, props, tree;

beforeEach(() => {
  props = createTestProps();
  wrapper = shallow(<Donation {...props} />);
});

describe("<Donation /> rendering", () => {
  it("Should render <Donation /> component", () => {
    expect(wrapper.exists()).toBe(true);
  });
});

describe("Display with all props (GBP) - image, name, date, comment, amount and GA amount", () => {
  it("Should contain a left column div", () => {
    expect(wrapper.find("div.donation_column_left")).toHaveLength(1);
  });

  it("Should have an img in the left column div", () => {
    expect(wrapper.find("div.donation_column_left img")).toHaveLength(1);
  });

  it("Should display image source from prop donor_avatar_url", () => {
    expect(wrapper.find("div.donation_column_left img").prop("src")).toEqual(
      DONOR_AVATAR_URL
    );
  });

  it("Should contain a right column div", () => {
    expect(wrapper.find("div.donation_column_right")).toHaveLength(1);
  });

  it("Should have a p.donation_date, in the right column div", () => {
    expect(
      wrapper.find("div.donation_column_right p.donation_date")
    ).toHaveLength(1);
  });

  it("Should have p.donation_date text matching donation_date prop", () => {
    expect(
      wrapper.find("div.donation_column_right p.donation_date").text()
    ).toEqual(DONATION_DATE_DISPLAY_FORMAT);
  });

  it("Should have a p.donor_name, in the right column div", () => {
    expect(wrapper.find("div.donation_column_right p.donor_name")).toHaveLength(
      1
    );
  });

  it("Should have p.donor_name text matching donor_name prop", () => {
    expect(
      wrapper.find("div.donation_column_right p.donor_name").text()
    ).toEqual(DONOR_NAME);
  });

  it("Should have a p.donor_name, in the right column div", () => {
    expect(wrapper.find("div.donation_column_right p.donor_name")).toHaveLength(
      1
    );
  });

  it("Should have p.donor_name text matching donor_name prop", () => {
    expect(
      wrapper.find("div.donation_column_right p.donor_name").text()
    ).toEqual(DONOR_NAME);
  });

  it("Should have a p.donor_comment, in the right column div", () => {
    expect(
      wrapper.find("div.donation_column_right p.donor_comment")
    ).toHaveLength(1);
  });

  it("Should have p.donor_comment text matching donor_comment prop", () => {
    expect(
      wrapper.find("div.donation_column_right p.donor_comment").text()
    ).toEqual(DONOR_COMMENT);
  });

  it("Should have a p.donation_amount, in the right column div", () => {
    expect(
      wrapper.find("div.donation_column_right p.donation_amount")
    ).toHaveLength(1);
  });

  it("Should have p.donation_amount text containing correctly formatted currency symbol and donation amount", () => {
    expect(
      wrapper
        .find("div.donation_column_right p.donation_amount")
        .childAt(0)
        .text()
    ).toEqual(DONATION_AMOUNT_CONVERTED);
  });

  it("Should have span.tax_reclaim within p.donation_amount", () => {
    expect(
      wrapper.find(
        "div.donation_column_right p.donation_amount span.tax_reclaim"
      )
    ).toHaveLength(1);
  });

  it("Should have span.tax_reclaim containing correctly formatted currency symbol and Gift Aid amount", () => {
    expect(
      wrapper
        .find("div.donation_column_right p.donation_amount span.tax_reclaim")
        .text()
    ).toEqual(DONATION_TAX_RECLAIM_CONVERTED);
  });

  it("Snapshot should display all data", () => {
    tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Display GBP donation amount correctly without Gift Aid", () => {
  it("Should have a p.donation_amount, in the right column div", () => {
    wrapper = shallow(
      <Donation {...props} donation_tax_reclaim={NO_TAX_RECLAIM} />
    );
    expect(
      wrapper.find("div.donation_column_right p.donation_amount")
    ).toHaveLength(1);
  });

  it("Should have p.donation_amount text containing correctly formatted currency symbol and donation amount", () => {
    wrapper = shallow(
      <Donation {...props} donation_tax_reclaim={NO_TAX_RECLAIM} />
    );
    expect(
      wrapper
        .find("div.donation_column_right p.donation_amount")
        .childAt(0)
        .text()
    ).toEqual(DONATION_AMOUNT_CONVERTED);
  });

  it("Should not have span within p.donation_amount", () => {
    wrapper = shallow(
      <Donation {...props} donation_tax_reclaim={NO_TAX_RECLAIM} />
    );
    expect(
      wrapper.find("div.donation_column_right p.donation_amount span")
    ).toHaveLength(0);
  });

  it("Snapshot should display donation without Gift Aid", () => {
    wrapper = shallow(
      <Donation {...props} donation_tax_reclaim={NO_TAX_RECLAIM} />
    );
    tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Display USD donation amount correctly", () => {
  it("Should have a p.donation_amount, in the right column div", () => {
    wrapper = shallow(
      <Donation
        {...props}
        donation_currency_local={USD_DONATION_CURRENCY_LOCAL}
        donation_amount_local={USD_DONATION_AMOUNT_LOCAL}
        donation_tax_reclaim={NO_TAX_RECLAIM}
      />
    );
    expect(
      wrapper.find("div.donation_column_right p.donation_amount")
    ).toHaveLength(1);
  });

  it("Should have p.donation_amount text containing correctly formatted currency symbol and donation amount", () => {
    wrapper = shallow(
      <Donation
        {...props}
        donation_currency_local={USD_DONATION_CURRENCY_LOCAL}
        donation_amount_local={USD_DONATION_AMOUNT_LOCAL}
        donation_tax_reclaim={NO_TAX_RECLAIM}
      />
    );
    expect(
      wrapper
        .find("div.donation_column_right p.donation_amount")
        .childAt(0)
        .text()
    ).toEqual(USD_DONATION_AMOUNT_CONVERTED);
  });

  it("Should not have span within p.donation_amount", () => {
    wrapper = shallow(
      <Donation
        {...props}
        donation_currency_local={USD_DONATION_CURRENCY_LOCAL}
        donation_amount_local={USD_DONATION_AMOUNT_LOCAL}
        donation_tax_reclaim={NO_TAX_RECLAIM}
      />
    );
    expect(
      wrapper.find("div.donation_column_right p.donation_amount span")
    ).toHaveLength(0);
  });

  it("Snapshot should display USD data", () => {
    wrapper = shallow(
      <Donation
        {...props}
        donation_currency_local={USD_DONATION_CURRENCY_LOCAL}
        donation_amount_local={USD_DONATION_AMOUNT_LOCAL}
        donation_tax_reclaim={NO_TAX_RECLAIM}
      />
    );
    tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
