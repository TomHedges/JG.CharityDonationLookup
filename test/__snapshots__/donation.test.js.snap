// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display GBP donation amount correctly without Gift Aid Snapshot should display donation without Gift Aid 1`] = `
<div
  className="donation_wrapper"
>
  <div
    className="donation_column_left"
  >
    <img
      className="donor_avatar"
      src="https://images.justgiving.com/image/f1e9879a-dace-443a-a453-d0636e66a348.jpg?template=profilesummary"
    />
  </div>
  <div
    className="donation_column_right"
  >
    <p
      className="donation_date"
    >
      01/04/2019
    </p>
    <p
      className="donor_name"
    >
      Tom Hedges
    </p>
    <p
      className="donor_comment"
    >
      Good luck!
    </p>
    <p
      className="donation_amount"
    >
      £10.00
    </p>
  </div>
</div>
`;

exports[`Display USD donation amount correctly Snapshot should display USD data 1`] = `
<div
  className="donation_wrapper"
>
  <div
    className="donation_column_left"
  >
    <img
      className="donor_avatar"
      src="https://images.justgiving.com/image/f1e9879a-dace-443a-a453-d0636e66a348.jpg?template=profilesummary"
    />
  </div>
  <div
    className="donation_column_right"
  >
    <p
      className="donation_date"
    >
      01/04/2019
    </p>
    <p
      className="donor_name"
    >
      Tom Hedges
    </p>
    <p
      className="donor_comment"
    >
      Good luck!
    </p>
    <p
      className="donation_amount"
    >
      $15.00
    </p>
  </div>
</div>
`;

exports[`Display with all props (GBP) should display image, name, date, comment, amount and GA amount Snapshot should display all data 1`] = `
<div
  className="donation_wrapper"
>
  <div
    className="donation_column_left"
  >
    <img
      className="donor_avatar"
      src="https://images.justgiving.com/image/f1e9879a-dace-443a-a453-d0636e66a348.jpg?template=profilesummary"
    />
  </div>
  <div
    className="donation_column_right"
  >
    <p
      className="donation_date"
    >
      01/04/2019
    </p>
    <p
      className="donor_name"
    >
      Tom Hedges
    </p>
    <p
      className="donor_comment"
    >
      Good luck!
    </p>
    <p
      className="donation_amount"
    >
      £10.00
      <span
        className="tax_reclaim"
      >
        + 
        £
        2.50
         Gift Aid
      </span>
    </p>
  </div>
</div>
`;
